libs/go/libp2p_node/aea/api.go

```diff
@@ -388,32 +399,80 @@ func (aea *AeaApi) listen_for_envelopes() {
 	}
 }
 
+func (aea *AeaApi) envelopeSendLoop() {
+	logger.Debug().Msg("send loop started")
+	var err error
+	for {
+		envelope := <-aea.send_queue
+		if envelope == nil {
+			logger.Info().Msg("envelope is nil. exit send loop")
+			return
+		}
+		err = aea.SendEnvelope(envelope)
+		if err != nil {
+			logger.Error().Str("err", err.Error()).Msg("while sending envelope")
+		} else {
+			logger.Debug().Msg("envelope sent")
+		}
+
+		if aea.closing {
+			return
+		}
+	}
+}
 func (aea *AeaApi) stop() {
-	aea.pipe.Close()
+	err := aea.pipe.Close()
+	if err != nil {
+		logger.Error().Str("err", err.Error()).Msgf("on pipe close during aeaapi stop")
+	}
 }
 
 /*
 
   Pipes helpers
 
 */
+const CurrentVersion = "0.1.0"
+
+func MakeACNMessageFromEnvelope(envelope *Envelope) (error, []byte) {
+	envelope_bytes, err := proto.Marshal(envelope)
+	if err != nil {
+		return err, envelope_bytes
+	}
+	return acn.EncodeAcnEnvelope(envelope_bytes)
+}
 
-func write_envelope(pipe Pipe, envelope *Envelope) error {
-	data, err := proto.Marshal(envelope)
+func (aea AeaApi) SendEnvelope(envelope *Envelope) error {
+	err, data := MakeACNMessageFromEnvelope(envelope)
 	if err != nil {
 		logger.Error().Str("err", err.Error()).Msgf("while serializing envelope: %s", envelope)
 		return err
 	}
-	return pipe.Write(data)
-}
+	err = aea.pipe.Write(data)
+	if err != nil {
+		logger.Error().Str("err", err.Error()).Msgf("on pipe write")
+		return err
+	}
+
+	status, err := acn.WaitForStatus(aea.acn_status_chan, AcnStatusTimeout)
 
-func read_envelope(pipe Pipe) (*Envelope, error) {
-	envelope := &Envelope{}
-	data, err := pipe.Read()
 	if err != nil {
-		logger.Error().Str("err", err.Error()).Msg("while receiving data")
-		return envelope, err
+		logger.Error().Str("err", err.Error()).Msgf("on status wait")
+		return err
 	}
-	err = proto.Unmarshal(data, envelope)
-	return envelope, err
+	if status.Code != acn.Status_SUCCESS {
+		logger.Error().
+			Str("op", "send_envelope").
+			Msgf("acn confirmation status is not Status Success: %d", status.Code)
+		return fmt.Errorf(
+			"send envelope: acn confirmation status is not Status Success: %d",
+			status.Code,
+		)
+	}
+	return err
```

Comment:
@solarw so on exit we don't retransmit the envelope? Even if it is not a connection error?

Reply:
yes, no idea what should do in this case. try to resend? how many times? once? twice?
i suppose in this case it's protocol issue: inproper version or encoding, so can not be fixed wih resending